modules/fx/arrow-fx-kotlinx-coroutines/src/test/kotlin/arrow/integrations/kotlinx/CoroutinesIntegrationTest.kt

```diff
@@ -0,0 +1,166 @@
+
+  init {
+    // --------------- suspendCancellable ---------------
+
+    "suspendedCancellable should throw" {
+      forAll(Gen.throwable()) { e ->
+        val ceh = TestCoroutineExceptionHandler()
+        val scope = TestCoroutineScope(ceh + TestCoroutineDispatcher())
+
+        scope.launch {
+          IO { throw e }.suspendCancellable()
+        }
+        // suspendCancellableCoroutine re throws the exception so we need to compare the instance
+        ceh.uncaughtExceptions[0]::class == e::class
+      }
+    }
+
+    "suspendedCancellable should resume with right block" {
+      forAll(Gen.int()) { i ->
+        val ceh = TestCoroutineExceptionHandler()
+        val scope = TestCoroutineScope(ceh + TestCoroutineDispatcher())
+        scope.launch {
+          val first = IO { i + 1 }.suspendCancellable()
+          val second = IO { first + 1 }.suspendCancellable()
+          val third = IO { second + 1 }.suspendCancellable()
+
+          third shouldBe i + 3
+        }
+
+        ceh.uncaughtExceptions.isEmpty()
+      }
+    }
+
+    "scope cancellation should cancel suspendedCancellable IO" {
+      forAll(Gen.int()) { i ->
+        IO.fx {
+          val scope = TestCoroutineScope(Job() + TestCoroutineDispatcher())
+          val promise = !Promise<Int>()
+          !effect {
+            scope.launch {
+              IO.cancelable<Unit> { promise.complete(i) }.suspendCancellable()
+            }
+          }
+          !effect { scope.cancel() }
+          !promise.get()
+        }.unsafeRunSync() == i
+      }
+    }
+
+    "suspendCancellable can cancel even for infinite asyncs" {
+      IO.async { cb: (Either<Throwable, Int>) -> Unit ->
+        val scope = TestCoroutineScope(Job() + TestCoroutineDispatcher())
+        scope.launch {
+          IO.async<Int> { Thread.sleep(5000) }
+            .onCancel(IO { cb(1.right()) })
+            .suspendCancellable().let {
+              cb(it.right())
+            }
+        }
+        IO(other) { Thread.sleep(500) }
+          .unsafeRunAsync { scope.cancel() }
+      }.unsafeRunTimed(2.seconds) shouldBe 1.some()
+    }
+
+    // --------------- unsafeRunScoped ---------------
+
+    "should rethrow exceptions within run block with unsafeRunScoped" {
+      forAll(Gen.throwable()) { e ->
+        try {
+          val scope = TestCoroutineScope(TestCoroutineDispatcher())
+          val ioa = IO<Int> { throw e }
+          ioa.unsafeRunScoped(scope) { either ->
+            either.fold({ throw it }, { fail("") })
+          }
+          fail("Should rethrow the exception")
+        } catch (throwable: Throwable) {
+          throwable == e
+        }
+      }
+    }
+
+    "unsafeRunScoped should cancel correctly" {
+      forAll(1, Gen.int()) { i ->
+        IO.async<Int> { cb ->
+          val scope = TestCoroutineScope(Job() + TestCoroutineDispatcher())
+          IO(all) { }
+            .flatMap { IO.async<Int> { cb -> Thread.sleep(200); cb(i.right()) } }
```

Comment:
```suggestion
            .flatMap { IO.asyncF<Int> { cb -> IO.sleep(200.milliseconds).followedBy(IO { cb(i.right()) }) } }
```

You can use `asyncF` to write an `IO` program within `async`, i.e. useful in cases where you need parallel, or concurrent ops.